src/inmanta/agent/agent.py

```diff
@@ -313,13 +313,23 @@ def __init__(self, agent, env_id, name, cache, ratelimiter):
         self.ratelimiter = ratelimiter
         self.version = 0
 
+    def get_scheduled_resource_actions(self):
+        return list(self.generation.values())
+
+    def finished(self):
+        for resource_action in self.generation.values():
+            if not resource_action.is_done():
+                return False
+        return True
+
     def reload(self, resources, undeployable={}, reason: str="RELOAD"):
         version = resources[0].id.get_version
 
         self.version = version
 
         for ra in self.generation.values():
             ra.cancel()
+        # TODO: Should we yield here to ensure that all running tasks are finished???
```

Comment:
yielding on all the futures for all the resource actions would ensure all tasks are finished.

However, there is no hard cancellation. Handlers that are executing are not interrupted, but continue to run.
In case of a handler being stuck, waiting here would make the agent stuck, as there is no way to start a new deploy.
Cancel does ensure that no new tasks are started. 

A wait with timeout would work, but I would test that very carefully, as we have only limited experience with the new timeout mechanisms of asyncio. What that timeout should be I can not say. 

The whole cancellation mechanism is something we should review, as asyncio now natively supports cancellation and we would like to have event propagation even in case of cancellation, but that is not an issue we should handle in this PR.

Reply:
Thanks for the clarification. Indeed, we should not handle this issue in this PR.